### PR TITLE
feat(windows): crash handler that writes minidumps to ...

### LINKED ISSUE #3111: Windows crash logs
Investigate adding Windows crash logs support similar to how we do it for Apple and Android (Microsoft-supported)

### PR DESCRIPTION
**Pull request by ReactorScram**
... ProgramData/dev.firezone.client/dumps/last_crash.dmp, in debug builds only for now

For #3111.
I was way off on my estimated dump size. It's only about 500 KB.

Please double-check that the Linux exes are still stripped properly. Windows, Linux, and Mac have different ways to do embedded or non-embedded debug info, so the Windows client should still be 20 MB but with a ~62MB pdb next to it, and the Linux client should still be 15 MB.

- [ ] Check Linux artifacts weren't accidentally affected
- [ ] Make sure connlib's panic handler doesn't interfere with this
- [ ] Later on, change CI so pdb files are saved as artifacts
- [ ] Later on, enable for release builds

### CODE REVIEW

File: rust/windows-client/src-tauri/Cargo.toml
Comment (jamilbk) [inline rust/windows-client/src-tauri/Cargo.toml:47]: ?? 🤣

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (jamilbk) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: I'm probably missing something, but why would this be relevant inside this file for the windows client? Is it called from the linux / android codepaths?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (jamilbk) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: It might be nice to log the filepath here so our more intrepid users can inspect as they wish.

File: rust/windows-client/src-tauri/src/client/gui.rs
Comment (jamilbk) [inline rust/windows-client/src-tauri/src/client/gui.rs]: (nit) This is called the bundle ID in apple or package ID in android. Might be nice to have a greppable constant name to be able to update these (we should really purchase our .com)

```suggestion
pub const BUNDLE_ID: &str = "dev.firezone.client";
```

File: rust/windows-client/src-tauri/src/client/gui.rs
Comment (jamilbk) [inline rust/windows-client/src-tauri/src/client/gui.rs]: I'm not the world's foremost expert in writing Rust unit tests, but would it make sense to modify the function signature for `run` to include a `segfault: bool` that we can then write a unit test for this?

Just trying to keep test codepaths out of the release codepaths (if it's not too cumbersome)

**Review by jamilbk — COMMENTED** [review]
LGTM, just some minor suggestions and a couple questions.

I don't feel comfortable approving, so I'll defer to @conectado to approve when he's back Monday.

File: rust/windows-client/src-tauri/Cargo.toml
Comment (ReactorScram) [reply rust/windows-client/src-tauri/Cargo.toml:47]: It's for simulating crashes lol
It doesn't have to go to production but if anyone else needs to check my work, it lets us do segfaults and other stuff that should trigger the crash handler

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: Yeah good point. I just copied it from the sample code.
If we generalize the Tauri code to make a Linux GUI later it might be handy?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: This happens in the crash handler process, so unfortunately these tracing logs don't actually go anywhere useful I think. I checked and even this "wrote minidump" isn't in the logs.

It might be hard for the crash handler process to append to the same connlib log file that the main process is using, too.

I was thinking of having the main process detect non-graceful shutdowns so even if the crash handler doesn't trigger, it can have a warning saying "Hey I didn't shut down gracefully last time, maybe it's a crash or maybe Task Manager killed me, check this path if you want"

File: rust/windows-client/src-tauri/src/client/gui.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/gui.rs]: Yeah I'm in favor of those kind of things. We had some discussion about that before with the `inject_faults` boolean that would intentionally slow down file I/O to make sure the GUI doesn't hang. It's hard to test things like this without having some "cheat code" paths to break things on purpose.

We can also test it externally by launching the GUI as a subprocess of a unit test and then sending a kill signal to it, maybe. I'm not sure if that will trigger the crash handler on Windows. 🤔

**Comment by ReactorScram** [comment]
Edit: Fixed. Good reason to keep it disabled in prod for now though.

Crud, it leaked a process somehow
I'll change this back to a draft, process leaks would be a disaster.

image

File: rust/windows-client/src-tauri/src/client/gui.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/gui.rs]: Yeah it's too easy to accidentally check in an `if true` there. I added a `--crash-on-purpose` flag.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: Why there's no handler for release?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: This shouldn't be ever necessary, what kind of complain?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: the `connect` naming here feels too vague

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: I'd rather use the module path explicitly, `Client` and `Server` are too vague but `minidump::Client` and `minidump::Server` is easy to read.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs:68]: why do it 10 times?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: What happens if we remove it?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: I think it's better if we just allow unsafe code generally and we should add a `SAFETY` comment here.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: we can remove this then

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: why is this needed?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: do we need to specify the type here?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs:33]: This is what sends the stack right?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: I'd rather add linux specific code paths when we implement the linux client and not preemptively

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs:70]: could we set a random socket name so we never overlap?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/crash_handling.rs]: rather than defining it inside a function let's do a submodule?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: Are these `tracing` calls going anywhere? I don't see any subscriber.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: I think Jamil wanted it disabled in Release mode

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: It would say that `attach_handler_inner` is dead code since the release builds wouldn't call it

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: Oops that was a copy-paste from the sample code. Removing it now

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs:33]: Yes, it sends stack traces for every thread and some basis system info like CPU type and I think maybe kernel version. https://docs.rs/minidumper/latest/minidumper/struct.Client.html#method.request_dump

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: They don't go anywhere. I assume it won't work to have two separate processes or subscribers writing to the same log files, but maybe I could set up a separate log dir for the crash handler process?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: Yeah. Changing that to `start_server_and_connect`

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs:68]: The sample code didn't say. Maybe in case the user tries to start a new instance of the application while the original is still writing its crash dump or something?

The sample code actually had an infinite loop, but I don't like infinite loops, so I limited it to 10.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs:70]: We could, but the client is already supposed to be single-instance because it always has the same name for the wintun adapter, same file path for the device ID, etc.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: I extracted it from the function, but the whole crash handling module is only 165 lines so I didn't split it yet.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: @jamilbk Is that right? Disable the crash handler in release mode? I don't remember what the conclusion was

Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: Jamil said that crash dumps should be enabled in Release builds if the dumps are moved into the log dir and exported when "Export Logs" is clicked.

If I do the subprocess mitigation for #2975  there will actually be two dump files probably, one for GUI crashes and one for connlib crashes, and 3 log dirs (crash handler, GUI process, connlib process), and those 5 things will all go into the zip archive.

Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: Opened #3263 to do that some other time.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: rather than this then we should `#[allow(dead_code)]` for for `attach_handler_inner` or maybe just don't have an inner one?

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (conectado) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs:70]: but if we do can we remove this loop?

**Review by conectado — APPROVED** [review]
LGTM! left a few responses but nothing blocking

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs]: I pushed a commit just now to try to remove the inner function.

I wanted to keep the `target_os` marker there even though the entire client is stubbed out on non-Windows platforms, just so we know the code will compile on other platforms but not work as intended.

And since it's stubbed out, it actually should be fine if this specific module doesn't even compile on other platforms.

File: rust/windows-client/src-tauri/src/client/crash_handling.rs
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/crash_handling.rs:70]: I added a comment saying that basically I don't know why it's there so I don't know if it's okay to remove it

**Pull request merged**
